docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+# RFC 26: Banning peers based on ResponseCheckTx
+
+## Changelog
+- Nov 4, 2022: Initial draft (jmalicevic)
+- Nov 8, 2022: Updated draft (jmalicevic)
+
+## Abstract
+
+Before adding a transaction to its mempool, a node runs `CheckTx` on it. `CheckTx` returns a non-zero code
+in case the transaction fails an application specific check. 
```

Comment:
To the best of my understanding, there are two ways in which `CheckTx` interferes here, and we should make this explicit:
1. no correct peer should ever add a transaction to the mempool (and gossip it) if it fails `CheckTx`
2. if the receiver of a transaction runs `CheckTx` and it fails, we assume the sender didn't do 1. and thus we may safely say that it is faulty.

The write-up focuses mostly on Point 2, but doesn't say explicitly what we expect to be the correct behavior of a peers that adds a transaction to the mempool. I think this might be not as obvious as it seems...

Reply:
The valid reason why 2 can happen is that  for example a node has gossiped a transaction that was valid at one height. Right after gossiping, the node commits a block, runs RecheckTx and the transaction is now invalid due to a changed state. The gossiped transaction just became invalid. Currently, outside of these cases, there should not be a way for an invalid transaction to be part of the mempool (it can be in the cache, but not mempool). 

But regardless of whether 2 can happen in the case a node is honest, you are right, we should explicitly define the correct behaviour.

Reply:
When working on @josef-widder's conditions above, also think of the case when a full node is a few heights late. It still keeps gossiping transactions, and so will run `CheckTx` on an "outdated" state w.r.t to its neighbor nodes.

Reply:
But in that case do we still want it to gossip nodes to us, or would we rather ban the node for a certain amount of time and then let it gossip transaction to us. Although , if we do ban it in this case, we might prevent it from reach our height (because we stopped gossiping to the node). Right now peers do not gossip transactions if the state of the peer they want to gossip to is more then one height away. This could be perhaps considered when banning as well? Although how many heights is enough is not deterministic?

Reply:
I would add to this discussion that it's also a piece of [node-local config](https://github.com/tendermint/tendermint/blob/bc59e16c533df793edb74672cc7b78d3600d9fc1/config/config.go#L713) whether or not re-checks occur. The effect of this for our purpose is that two peers A and B may be at the same height and therefore gossiping transactions to each other. However, if A does not have recheck enabled, it may gossip some transaction to B that was made invalid by a recent height but that it didn't clear from the mempool. I'm not certain I'd categorize having recheck disabled as misconfiguration. It means that we may receive invalidated Txs from otherwise well-behaved nodes that have reheck disabled.